core/src/Borrower.sol

```diff
@@ -70,26 +84,56 @@ contract Borrower is IUniswapV3MintCallback {
         packedSlot.owner = owner;
     }
 
-    // TODO liquidations
-    function liquidate() external {
+    function liquidate(ILiquidator callee, bytes calldata data) external {
         require(!packedSlot.isInCallback);
 
-        int24[] memory positions_ = positions.read();
-
-        (, int24 currentTick, , , , , ) = UNISWAP_POOL.slot0();
         (uint256 liabilities0, uint256 liabilities1) = _getLiabilities();
         Prices memory prices = _getPrices();
-        Assets memory assets = _getAssets(
-            positions_,
-            Uniswap.FeeComputationCache(
-                currentTick,
-                UNISWAP_POOL.feeGrowthGlobal0X128(),
-                UNISWAP_POOL.feeGrowthGlobal1X128()
-            ),
-            prices
-        );
+        Assets memory assets = _getAssetsAndWithdrawPositions(positions.read(), prices);
+
+        require(!BalanceSheet.isHealthy(liabilities0, liabilities1, assets, prices));
+
+        uint256 assets0 = TOKEN0.balanceOf(address(this));
+        uint256 repayable0 = Math.min(assets0, liabilities0);
+        liabilities0 -= repayable0;
+
+        uint256 assets1 = TOKEN1.balanceOf(address(this));
+        uint256 repayable1 = Math.min(assets1, liabilities1);
+        liabilities1 -= repayable1;
+
+        // TODO: this is already computed inside of computeLiquidationIncentive, so use that instead of
+        // re-computing it (or at least compare gas between the 2 options)
+        uint256 priceX96 = Math.mulDiv(prices.c, prices.c, FixedPoint96.Q96);
 
-        if (!BalanceSheet.isHealthy(liabilities0, liabilities1, assets, prices)) packedSlot.owner = msg.sender;
+        if (liabilities0 + liabilities1 == 0 || (liabilities0 > 0 && liabilities1 > 0)) {
+            // If both are zero or neither is zero, there's nothing more to do
+            // TODO: compensate liquidators for txn costs using ANTE
+            return;
+        } else if (liabilities0 > 0) {
```

Comment:
For any remaining liabilities, we _do_ need the liquidator to get more involved. They have to swap the asset that's in surplus for the asset that's lacking.